### PR TITLE
LibIPC+Everywhere: Introduce an IPC Transport abstraction 

### PR DESCRIPTION
**Pull request by ADKaster**
This abstraction will help us to support multiple IPC transport
mechanisms going forward. For now, we only have a TransportSocket that
implements the same behavior we previously had, using Unix Sockets for
IPC.

As part of this refactor,  move Core::IPCProcess into WebView::Process.
Ladybird's HelperProcess.cpp was the only user of the IPCProcess class.
Moving the helper class from LibCore allows for some more interesting
LibIPC changes in the upcoming commits.

### CODE REVIEW

File: Userland/Libraries/LibWeb/HTML/MessagePort.cpp [hunk 1]
```diff
@@ -320,18 +335,13 @@ ErrorOr<MessagePort::ParseDecision> MessagePort::parse_message()
...
-    u8 buffer[4096] {};
-
-    Vector<int> fds;
-    // FIXME: What if pending bytes is > 4096? Should we loop here?
-    auto maybe_bytes = m_socket->receive_message(buffer, MSG_NOSIGNAL, fds);
-    if (maybe_bytes.is_error()) {
-        dbgln("MessagePort::read_from_socket(): Failed to receive message: {}", maybe_bytes.error());
-        return;
-    }
-    auto bytes = maybe_bytes.release_value();
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/HTML/MessagePort.cpp:334]: the "should we loop" fixme is now resolved by calling the same API that IPC does. the "perform_shutdown" callback is a bit iffy, but I found that without the global task, we ended up in the "function being destroyed while executing" invalid state of AK::Function.

**Review by ADKaster — COMMENTED** [review]
Couple self-comments. Let me know which of these I should fix before merge and which deserve an issue.

File: Userland/Libraries/LibIPC/ConnectionToServer.h [hunk 2]
```diff
@@ -11,18 +11,18 @@
...
-        TRY(socket->set_blocking(true));                                                               \
-                                                                                                       \
-        return adopt_nonnull_ref_or_enomem(new (nothrow) Klass(move(socket), forward<Args>(args)...)); \
+#define IPC_CLIENT_CONNECTION(klass, socket_path)                                                                      \
+    C_OBJECT_ABSTRACT(klass)                                                                                           \
+public:                                                                                                                \
+    template<typename Klass = klass, class... Args>                                                                    \
+    static ErrorOr<NonnullRefPtr<klass>> try_create(Args&&... args)                                                    \
+    {                                                                                                                  \
+        auto parsed_socket_path = TRY(Core::SessionManagement::parse_path_with_sid(socket_path));                      \
```
[the remark this answers is not in the document]
Comment (awesomekling) [reply Userland/Libraries/LibIPC/ConnectionToServer.h:20]: If we're not using it in this codebase, it's fine to remove it here :)

File: Userland/Libraries/LibIPC/TransportSocket.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        if (maybe_nwritten.is_error()) {
+            if (auto error = maybe_nwritten.release_error(); error.is_errno() && (error.code() == EAGAIN || error.code() == EWOULDBLOCK)) {
+                Vector<struct pollfd, 1> pollfds;
+                if (pollfds.is_empty())
+                    pollfds.append({ .fd = m_socket->fd().value(), .events = POLLOUT, .revents = 0 });
+
+                ErrorOr<int> result { 0 };
+                do {
+                    constexpr u32 POLL_TIMEOUT_MS = 100;
+                    result = Core::System::poll(pollfds, POLL_TIMEOUT_MS);
```
[the remark this answers is not in the document]
Comment (trflynn89) [reply Userland/Libraries/LibIPC/TransportSocket.cpp:75]: Yeah the latter sounds doable - if we can only send a partial message, makes sense to put the remainder of the message back in front queue and try again later.

File: Userland/Libraries/LibIPC/TransportSocket.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    void wait_until_readable();
+
+    ErrorOr<void> transfer(ReadonlyBytes, Vector<int, 1> const& unowned_fds);
+
+    struct [[nodiscard]] ReadResult {
+        Vector<u8> bytes;
+        Vector<int> fds;
+    };
+    [[nodiscard]] ReadResult read_as_much_as_possible_without_blocking(Function<void()> schedule_shutdown);
```
Comment (trflynn89) [inline Userland/Libraries/LibIPC/TransportSocket.h:33]: Don't think we need both the type and the method to have `[[nodiscard]]`

**Review by trflynn89 — COMMENTED** [review]
Looks reasonable to me - agreed it's fine to just remove Serenity-only APIs

File: Userland/Libraries/LibIPC/TransportSocket.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ADKaster) [reply Userland/Libraries/LibIPC/TransportSocket.cpp:75]: Sounds good. going to add a FIXME for putting the partial write back onto the queue until the socket is writeable.

**Pull request merged**